[libc++] Optionally support filecheck and split-file

ldionne opened the pull request:
This patch adds optional support for FileCheck and split-file in libc++'s test 
suite. Whether FileCheck / split-file is available can be queried using Lit
features.

This should make it possible to test several things that were previously
impossible to test, especially for specific code generation.

Supersedes #65917 and #188283

ldionne committed:
[libc++] Optionally support filecheck-based tests

This patch adds support for filecheck tests in libc++'s test suite.
However, it doesn't use LLVM's filecheck program, which requires
building a bunch of LLVM utilities. Instead, it installs the Python
port of filecheck at https://github.com/AntonLydike/filecheck which
supports basically the same functionality.

The test suite still works when filecheck is not available, since
tests that use filecheck should be guarded on the `has-filecheck`
Lit feature.

This should make it possible to test several things that were
previously impossible to test, especially for specific code generation.

boomanaiden154 reviewed (commented):
Drive by comment from a non libc++ contributor, so take my feedback as you will.

Why install a Python version of FileCheck when we could just install/build actual FileCheck? I'd prefer to avoid external dependencies whenever possible, especially when it's a much less battle tested reimplementation of something that we have in tree.

ldionne commented:
> Drive by comment from a non libc++ contributor, so take my feedback as you will.
> 
> Why install a Python version of FileCheck when we could just install/build actual FileCheck? I'd prefer to avoid external dependencies whenever possible, especially when it's a much less battle tested reimplementation of something that we have in tree.

The problem is that AFAICT, LLVM's `FileCheck` requires configuring the LLVM project as a whole and building a ton of things we don't need for the runtimes. I wouldn't even be surprised if it didn't build on some smaller platforms that we target.

philnik777 commented on libcxx/utils/libcxx/test/features.py:
```diff
@@ -355,6 +355,14 @@ def _mingwSupportsModules(cfg):
+    Feature(
+        name="has-filecheck",
+        when=lambda cfg: runScriptExitCode(cfg, ["filecheck --version"]) == 0,
+    ),
```
I think allowing the external version is very reasonable, but IMO we should also look for `FileCheck`, since folks may very well have that already on their system (e.g. I do).

ldionne replied:
The issue though is that we have to standardize on the case we're going to use in the tests. LLVM uses `FileCheck` and the Python port is `filecheck` (all lowercase).

boomanaiden154 commented:
> so you'd basically be looking at doubling the build time. It's not definitely unreasoable, since libc++ is relatively quick to build, but still a lot worse than 188 steps might sound at first. Anyways, I think the much bigger problem would be that we'd create a circular dependency between the runtimes build and the llvm build.

It looks like the plan here is to find `FileCheck` (or a python equivalent) as a system dependency though? I think that should alleviate any concerns with circular dependencies/build times given it needs to only happen at CI image build time. Given that doesn't happen very often, I don't see how build times would be a concern either.

ldionne commented:
> It looks like the plan here is to find `FileCheck` (or a python equivalent) as a system dependency though? I think that should alleviate any concerns with circular dependencies/build times given it needs to only happen at CI image build time. Given that doesn't happen very often, I don't see how build times would be a concern either.

If a user wants to set up `FileCheck` locally to work on libc++, they shouldn't have to configure LLVM and build it. It's a fragile and complicated process. With this patch, they only need to do `pip install filecheck` and they're done.

If we productized `FileCheck` into something standalone that can be installed with e.g. Homebrew, I'd much prefer use the LLVM version. But since I don't see that happening, I think it makes more sense to use an easily installable equivalent of it.

boomanaiden154 commented:
Another option: We could do what compiler-rt does and only use FileCheck when doing a bootstrapping build (`LLVM_ENABLE_PROJECTS="clang" -DLLVM_ENABLE_RUNTIMES="libcxx" ../llvm`). That solves two problems:
1. We can get `FileCheck` from the LLVM build.
2. We ensure the tooling that we're testing with these tests (the compiler/diagnostic tooling) is versioned atomically with libc++ which I think would remove some churn.

That might make it harder for some platforms to run these specific tests. I would imagine that just checking diagnostics/codegen is something where we do not care too much about running everywhere because we're more of testing the hosted tools, but not sure what the exact intentions are.

ldionne committed:
Merge branch 'main' into review/add-optional-filecheck

# Conflicts:
#	.github/workflows/libcxx-build-and-test.yaml
#	libcxx/utils/ci/Dockerfile
#	libcxx/utils/libcxx/test/features.py

ldionne committed:
Support both filecheck and FileCheck

ldionne committed:
Fix selftest

ldionne committed:
Add negative filecheck test

ldionne committed:
Don't install in the Docker image

Long term, this is something that should be done automatically by the
test suite on setup.

ldionne committed:
Try installing venv on Linux runners

ldionne committed:
sudo

ldionne committed:
Using apt-get is harder than it looks

ldionne committed:
format

Michael137 reviewed (approved):
Tested this locally and it works great with the tests I'm about to put up. So LGTM

I'll let the libc++/CI maintainers give final say

ldionne committed:
Merge branch 'main' into review/add-optional-filecheck

boomanaiden154 reviewed (commented):
I've just setup https://pypi.org/project/llvm-testing-tools which should contain `FileCheck` and `split-file`. Can we look at using that?

Not sure what kind of platform support we want for FileCheck based tests for now. That package only currently has supported for x86 Linux with glibc 2.39+. I can add more support as needed though.

ldionne committed:
Merge branch 'main' into review/add-optional-filecheck

ldionne committed:
Try using Pip package for llvm-testing-utils

ldionne commented:
> I've just setup https://pypi.org/project/llvm-testing-tools which should contain `FileCheck` and `split-file`. Can we look at using that?
> 
> Not sure what kind of platform support we want for FileCheck based tests for now. That package only currently has supported for x86 Linux with glibc 2.39+. I can add more support as needed though.

Thanks for doing that. I just updated the patch to install `llvm-testing-tools` instead. I think we'll need this on all the platforms that we test on, which might be annoying to setup (Windows, macOS, Linux). That's also why a pure-Python package is simpler.

Do you think it would be feasible to provide this package on Windows and macOS as well?

ldionne committed:
Update libcxx/utils/libcxx/test/features/misc.py

ldionne committed:
Pin versions of libc++ testing requirements

boomanaiden154 reviewed (approved):
One minor nit, otherwise LGTM.

Thanks for doing this. I'm excited to see the testing improvements this can bring to libc++. Sorry for all the run around on approach. I wasn't trying to make this more difficult but probably did.

ldionne committed:
Use not instead of XFAIL
